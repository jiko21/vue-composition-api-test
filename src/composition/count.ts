import { ref } from '@vue/composition-api';

const useCount = () => {
  const countValue = ref(0);
  const increment = () => {
    countValue.value += 1;
  };
  const decrement = () => {
    countValue.value -= 1;
  };
  return {
    countValue,
    increment,
    decrement,
  };
};

// eslint-disable-next-line import/prefer-default-export
export { useCount };
