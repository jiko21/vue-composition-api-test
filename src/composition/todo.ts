import { computed, reactive } from '@vue/composition-api';

const useTodo = () => {
  const todo = reactive({
    todos: [] as string[],
    length: computed(() => todo.todos.length),
  }) as any;
  const addTodo = (item: string) => {
    todo.todos.push(item);
  };
  const deleteTodo = (index: number) => {
    todo.todos.splice(index, 1);
  };
  return {
    todo,
    addTodo,
    deleteTodo,
  };
};

// eslint-disable-next-line import/prefer-default-export
export { useTodo };
