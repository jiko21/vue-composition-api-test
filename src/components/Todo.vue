<template>
  <div class="count">
    <input id="todo-input" v-model="text"/>
    <button class="add-btn" @click="onSubmit">追加</button>
    <ul>
      <li v-for="(task, i) in todo.todos" :key="i">
        <p>{{task}}</p>
        <button class="delete-btn" @click="deleteTodo(i)">Delete</button>
      </li>
    </ul>
  </div>
</template>

<script lang="ts">
import { createComponent, ref } from '@vue/composition-api';
import { useTodo } from '@/composition/todo';

export default createComponent({
  name: 'Todo',
  setup() {
    const text = ref('');
    const { todo, addTodo, deleteTodo } = useTodo();
    const onSubmit = () => {
      addTodo(text.value);
      text.value = '';
    };
    return {
      text,
      onSubmit,
      todo,
      addTodo,
      deleteTodo,
    };
  },
});
</script>

<style scoped>
</style>
