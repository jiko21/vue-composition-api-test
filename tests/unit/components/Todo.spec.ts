import { createLocalVue, mount, shallowMount } from '@vue/test-utils';
import VueCompositionApi from '@vue/composition-api';
import Todo from '@/components/Todo.vue';
import * as composition from '@/composition/todo';


const localVue = createLocalVue();
localVue.use(VueCompositionApi);

let addTodoMock: jest.Mock;
let deleteTodoMock: jest.Mock;

describe('Todo.vue', () => {
  beforeEach(() => {
    jest.mock('@/composition/todo');
    addTodoMock = jest.fn();
    deleteTodoMock = jest.fn();
    const TODOS = [
      'アドベントカレンダー',
      '修論',
      '筋トレ',
    ];
    jest.spyOn(composition, 'useTodo').mockReturnValue({
      todo: {
        todos: TODOS,
        length: () => TODOS.length,
      },
      addTodo: addTodoMock,
      deleteTodo: deleteTodoMock,
    });
  });
  it('correctly renders initial html', () => {
    const wrapper = shallowMount(Todo, {
      localVue,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('correctly call addTodo when `追加` button is clicked', () => {
    const wrapper = shallowMount(Todo, {
      localVue,
    });
    wrapper.find('#todo-input').setValue('ポスターセッション');
    wrapper.find('.add-btn').trigger('click');
    expect(addTodoMock).toHaveBeenCalledWith('ポスターセッション');
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('correctly call deleteTodo when `Delete` button is clicked', () => {
    const wrapper = shallowMount(Todo, {
      localVue,
    });
    const INDEX = 1;
    wrapper.findAll('.delete-btn').at(INDEX).trigger('click');
    expect(deleteTodoMock).toHaveBeenCalledWith(INDEX);
  });
});
