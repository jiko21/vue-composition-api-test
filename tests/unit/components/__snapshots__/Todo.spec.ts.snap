// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo.vue correctly call addTodo when \`追加\` button is clicked 1`] = `
<div class="count"><input id="todo-input"> <button class="add-btn">追加</button>
  <ul>
    <li>
      <p>アドベントカレンダー</p> <button class="delete-btn">Delete</button>
    </li>
    <li>
      <p>修論</p> <button class="delete-btn">Delete</button>
    </li>
    <li>
      <p>筋トレ</p> <button class="delete-btn">Delete</button>
    </li>
  </ul>
</div>
`;

exports[`Todo.vue correctly renders initial html 1`] = `
<div class="count"><input id="todo-input"> <button class="add-btn">追加</button>
  <ul>
    <li>
      <p>アドベントカレンダー</p> <button class="delete-btn">Delete</button>
    </li>
    <li>
      <p>修論</p> <button class="delete-btn">Delete</button>
    </li>
    <li>
      <p>筋トレ</p> <button class="delete-btn">Delete</button>
    </li>
  </ul>
</div>
`;
